kernel/src/fs/overlayfs/fs.rs

```diff
@@ -0,0 +1,991 @@
+            // the rmdir will fail and the later create will fail.
+            let _ = upper.rmdir(name);
+        }
+
+        let new_upper = upper.create(name, type_, mode)?;
+
+        let new_child = Arc::new_cyclic(|weak| OverlayInode {
+            ino: self.overlay_fs().alloc_ino(),
+            type_,
+            name: String::from(name),
+            parent: Some(self.self_.upgrade().unwrap()),
+            upper: Mutex::new(Some(new_upper)),
+            lowers: Vec::new(),
+            fs: self.fs.clone(),
+            self_: weak.clone(),
+        });
+        Ok(new_child)
+    }
+
+    /// Writes data to the target inode, if it resides in the lower layer,
+    /// it will be copied up to the upper layer.
+    /// The corresponding parent directories will be created also if they do not exist.
+    pub fn write_at(&self, offset: usize, reader: &mut VmReader) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        let upper = self.build_upper_recursively_if_needed()?;
+        upper.write_at(offset, reader)
+    }
+
+    pub fn write_direct_at(&self, offset: usize, reader: &mut VmReader) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        let upper = self.build_upper_recursively_if_needed()?;
+        upper.write_direct_at(offset, reader)
+    }
+
+    pub fn read_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        self.get_top_valid_inode().read_at(offset, writer)
+    }
+
+    pub fn read_direct_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        self.get_top_valid_inode().read_direct_at(offset, writer)
+    }
+
+    /// Returns the children objects in a unified view.
+    /// The object from the upper layer with the same name will mask the lower ones.
+    pub fn readdir_at(&self, offset: usize, visitor: &mut dyn DirentVisitor) -> Result<usize> {
+        if self.type_ != InodeType::Dir {
+            return_errno!(Errno::ENOTDIR);
+        }
+
+        let mut overlay_dir_visitor = OverlayDirMergedVisitor::new();
+        if let Some(upper) = self.upper() {
+            debug_assert!(upper.type_() == InodeType::Dir);
+            // The upper cannot be an opaque.
+            upper.readdir_at(offset, &mut overlay_dir_visitor)?;
+        }
+
+        for lower in self.lowers.iter() {
+            debug_assert!(lower.type_() == InodeType::Dir);
+            // The lower cannot be an opaque.
+            lower.readdir_at(offset, &mut overlay_dir_visitor)?;
+        }
+
+        let mut cnt = 0;
+        for (offset, (name, ino, type_)) in overlay_dir_visitor.as_merged_view() {
+            visitor.visit(name, *ino, *type_, *offset)?;
```

Comment:
The `ino` of an OverlayFS inode returned from `readdir_at` is inconsistent with that from the `metadata` method.

In addition, the inode numbers from the underlying FSes are directly exposed to the user space. How can an instance of OverlayFS ensure the uniqueness of its node numbers?